syntax = "proto3";

import "google/protobuf/timestamp.proto";

package doca_flow_counter_spy;

service CounterSpy {
    rpc getFlowCounts(EmptyRequest) returns (QueryResult) {}
}

enum PortType {
    PIPE_BASIC = 0;
    PIPE_CONTROL = 1;
    PIPE_LPM = 2;
    PIPE_CT = 3;
    PIPE_ACL = 4;
    PIPE_ORDERED_LIST = 5;
    PIPE_HASH = 6;
}

message EmptyRequest {}

message Entry
{
    uint64 id = 1;
    uint64 total_packets = 2;
    uint64 total_bytes = 3;
    uint64 delta_packets = 4;
    uint64 delta_bytes = 5;
}

message Pipe
{
    string name = 1;
    repeated Entry entries = 2;
    repeated Entry shared_counters = 3;
    bool is_root = 4;
    PortType type = 5;
}

message Port
{
    uint32 port_id = 1;
    repeated Pipe pipes  = 2;
    repeated Entry shared_counters = 3;
}

message QueryResult
{
    google.protobuf.Timestamp timestamp = 1;
    repeated Port ports = 2;
}
